File: app/src/main/java/nl/rsdt/japp/jotial/maps/deelgebied/Deelgebied.java
package nl.rsdt.japp.jotial.maps.deelgebied;

import android.content.res.Resources;
import android.graphics.Color;
import android.location.Location;
import android.util.Log;

import com.google.android.gms.maps.model.LatLng;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.maps.android.PolyUtil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Objects;

import nl.rsdt.japp.R;
import nl.rsdt.japp.application.JappPreferences;
import nl.rsdt.japp.jotial.data.structures.area348.MetaColorInfo;
import nl.rsdt.japp.jotial.maps.kml.KmlDeelgebied;
import nl.rsdt.japp.jotial.maps.kml.KmlFile;
import nl.rsdt.japp.jotial.maps.kml.KmlLocation;
import nl.rsdt.japp.jotial.maps.kml.KmlReader;

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 13-7-2016
 * Class for the Deelgebieden,
 * Deelgebied.initialize(Resources) must be called in order for this class to function.
 */
public final class Deelgebied {

    /**
     * Defines the Alpha Deelgebied.
     * */
    private static final Deelgebied Alpha = new Deelgebied("alpha", JappPreferences.getColorName("a"));

    /**
     * Defines the Bravo Deelgebied.
     * */
    private static final Deelgebied Bravo = new Deelgebied("bravo", JappPreferences.getColorName("b"));

    /**
     * Defines the Charlie Deelgebied.
     * */
    private static final Deelgebied Charlie = new Deelgebied("charlie", JappPreferences.getColorName("c"));

    /**
     * Defines the Delta Deelgebied.
     * */
    private static final Deelgebied Delta = new Deelgebied("delta", JappPreferences.getColorName("d"));

    /**
     * Defines the Echo Deelgebied.
     * */
    private static final Deelgebied Echo = new Deelgebied("echo", JappPreferences.getColorName("e"));

    /**
     * Defines the Foxtrot Deelgebied.
     * */

    private static final Deelgebied Foxtrot = new Deelgebied("foxtrot", JappPreferences.getColorName("f"));

    /**
     * Defines the Xray Deelgebied.
     * */
    public static final Deelgebied Xray = new Deelgebied("xray", JappPreferences.getColorName("x"));
    private static final String TAG = "Deelgebied";

    /**
     * Gets a array of all the Deelgebieden.
     * */
    public static Deelgebied[] all() {
        return new Deelgebied[] { Alpha, Bravo, Charlie, Delta, Echo, Foxtrot, Xray };
    }
    public static volatile boolean deelgebiedenInitialized = false;

    /**
     * The name of this Deelgebied.
     * */
    private String name;

    /**
     * The hunt drawable associated with this Deelgebied.
     * */
    private int drawable_hunt;

    /**
     * The spot drawable associated with this Deelgebied.
     * */
    private int drawable_spot;

    /**
     * The color of this Deelgebied.
     * */
    private int color;

    /**
     * Gets the hunt drawable associated with this Deelgebied.
     * */
    public int getDrawableHunt() {
        return drawable_hunt;
    }

    /**
     * Gets the spot drawable associated with this Deelgebied.
     * */
    public int getDrawableSpot() {
        return drawable_spot;
    }

    /**
     * Gets the color of the Deelgebied.
     * */
    public int getColor() {
        return color;
    }

    /**
     * Gets the name of the Deelgebied.
     * */
    public String getName() {
        return name;
    }

    /**
     * Gets the color of the Deelgebied, with a given alpha.
     *
     * @param alpha The alpha from 0 - 255.
     * @return The color of the Deelgebied with a given alpha.
     * */
    public int alphaled(int alpha) {
        return Color.argb(alpha, Color.red(color), Color.green(color), Color.blue(color));
    }

    /**
     * The list of coordinates that is the area of the Deelgebied.
     * */
    private ArrayList<LatLng> coordinates = new ArrayList<>();
    private static Map<String,LinkedList<OnInitialized>> onInitializedList = new HashMap<>();
    /**
     * Gets the list of coordinates that is the area of the Deelgebied.
     * */
    public ArrayList<LatLng> getCoordinates() {
        return this.coordinates;
    }

    public void getDeelgebiedAsync(OnInitialized onInitialized){
        if (deelgebiedenInitialized){
            onInitialized.onInitialized(this);
        }
        else{
            if (!onInitializedList.containsKey(this.name)) onInitializedList.put(this.name, new LinkedList<>());
            onInitializedList.get(this.name).add(onInitialized);
        }
    }

    /**
     * Initializes a new instance of Deelgebied.
     *
     * @param name The name of the Deelgebied.
     * */
    private Deelgebied(String name,String colorName)
    {
        this.name = name;
        switch(MetaColorInfo.ColorNameInfo.DeelgebiedColor.valueOf(colorName)){

            case Groen:
                this.drawable_hunt = R.drawable.vos_groen_4;
                this.drawable_spot = R.drawable.vos_groen_3;
                this.color = Color.argb(255, 0, 255, 0);
                break;
            case Rood:
                this.drawable_hunt = R.drawable.vos_rood_4;
                this.drawable_spot = R.drawable.vos_rood_3;
                this.color = Color.argb(255, 255, 0, 0);
                break;
            case Paars:
                this.drawable_hunt = R.drawable.vos_paars_4;
                this.drawable_spot = R.drawable.vos_paars_3;
                this.color = Color.argb(255, 255, 0, 255);
                break;
            case Oranje:
                this.drawable_hunt = R.drawable.vos_oranje_4;
                this.drawable_spot = R.drawable.vos_oranje_3;
                this.color = Color.argb(255, 255, 162, 0);
                break;
            case Blauw:
                this.drawable_hunt = R.drawable.vos_oranje_4;
                this.drawable_spot = R.drawable.vos_oranje_3;
                this.color = Color.argb(255, 0, 0, 255);
                break;
            default:
            case Onbekend:
            case Zwart:
                this.drawable_hunt = R.drawable.vos_zwart_4;
                this.drawable_spot = R.drawable.vos_zwart_3;
                this.color = Color.argb(255, 0, 0, 0);
                break;
            case Turquoise:
                this.drawable_hunt = R.drawable.vos_groen_4;
                this.drawable_spot = R.drawable.vos_groen_3;
                this.color = Color.argb(255, 0, 255, 255);
                break;
        }

    }

    /**
     * Checks if the DeelgebiedData contains a location.
     *
     * @param location The location to check.
     * @return Value indicating if the location is within the DeelgebiedData.
     * */
    public boolean containsLocation(LatLng location)
    {
        return PolyUtil.containsLocation(location, getCoordinates(), false);
    }

    /**
     * Checks if the DeelgebiedData contains a location.
     *
     * @param location The location to check.
     * @return Value indicating if the location is within the DeelgebiedData.
     * */
    public boolean containsLocation(Location location)
    {
        return PolyUtil.containsLocation(new LatLng(location.getLatitude(), location.getLongitude()), getCoordinates(), false);
    }


    /**
     * Initializes the Deelgebieden, loading their coordinates from the Resources.
     * */
    public static synchronized void initialize(Resources resources)
    {
        KmlReader.parseFromMeta(new KmlReader.Callback() {
            @Override
            public void onException(Throwable e) {
                Log.e(Deelgebied.TAG,e.toString());
                Deelgebied[] gebieden = all();
                Deelgebied current;
                for(int g = 0; g < gebieden.length; g++) {
                    current = gebieden[g];

                    InputStream stream;
                    switch (current.name) {
                        case "alpha":
                            stream = resources.openRawResource(R.raw.alpha);
                            break;
                        case "bravo":
                            stream = resources.openRawResource(R.raw.bravo);
                            break;
                        case "charlie":
                            stream = resources.openRawResource(R.raw.charlie);
                            break;
                        case "delta":
                            stream = resources.openRawResource(R.raw.delta);
                            break;
                        case "echo":
                            stream = resources.openRawResource(R.raw.echo);
                            break;
                        //case "foxtrot":
                            //stream = resources.openRawResource(R.raw.foxtrot);
                            //break;
                        default:
                            stream = null;
                            Log.i("Deelgebied", "No polygon data was found for " + current.name);
                            break;
                    }

                    if (stream != null) {
                        BufferedReader r = new BufferedReader(new InputStreamReader(stream));
                        StringBuilder total = new StringBuilder();
                        String line;
                        try {
                            while ((line = r.readLine()) != null) {
                                total.append(line).append('\n');
                            }
                        } catch (IOException err) {
                            Log.e("Deelgebied", "Error occurred while reading stream", err);
                        }

                        String data = total.toString();

                        current.coordinates = new Gson().fromJson(data, new TypeToken<ArrayList<LatLng>>() {
                        }.getType());
                    }
                }
                deelgebiedenInitialized = true;
                for (Map.Entry<String, LinkedList<OnInitialized>> entry: onInitializedList.entrySet()){
                    for (OnInitialized oi : entry.getValue()){
                        oi.onInitialized(Objects.requireNonNull(Deelgebied.parse(entry.getKey())));
                    }
                }
            }

            @Override
            public void onSucces(KmlFile kml) {
                Deelgebied[] gebieden = all();
                Deelgebied current;
                for(int g = 0; g < gebieden.length; g++) {
                    current = gebieden[g];

                    KmlDeelgebied kmldg;
                    switch (current.name) {
                        case "alpha":
                            kmldg = kml.getAlpha();
                            break;
                        case "bravo":
                            kmldg = kml.getBravo();
                            break;
                        case "charlie":
                            kmldg = kml.getCharlie();
                            break;
                        case "delta":
                            kmldg = kml.getDelta();
                            break;
                        case "echo":
                            kmldg = kml.getEcho();
                            break;
                        case "foxtrot":
                            kmldg = kml.getFoxtrot();
                            break;
                        default:
                            kmldg = null;
                            break;
                    }

                    if (kmldg != null) {

                        ArrayList<LatLng> tmp = new ArrayList<>(kmldg.getBoundry().size());
                        for (int i = 0; i < kmldg.getBoundry().size(); i++) {
                            KmlLocation kmldgLoc = kmldg.getBoundry().get(i);
                            LatLng tmploc = new LatLng(kmldgLoc.lat, kmldgLoc.lon);
                            while (tmp.size()-1 < i){
                                tmp.add(null);
                            }
                            tmp.set(i, tmploc);
                        }
                        current.coordinates = tmp;
                    } else {
                        Log.i("Deelgebied", "No polygon data was found for " + current.name);
                    }
                }
                deelgebiedenInitialized = true;
                for (Map.Entry<String, LinkedList<OnInitialized>> entry: onInitializedList.entrySet()){
                    for (OnInitialized oi : entry.getValue()){
                        oi.onInitialized(Objects.requireNonNull(Deelgebied.parse(entry.getKey())));
                    }
                }
            }
        });
        waitUntilInitzialized();
    }

    /**
     * Resolves the Deelgebied on the location.
     *
     * @param location The location where the Deelgebied can be resolved on.
     * @return The Deelgebied resolved from the location, returns null if none.
     */
    public static Deelgebied resolveOnLocation(LatLng location)
    {
        waitUntilInitzialized();
        Deelgebied[] data = all();
        Deelgebied current;

        for(int i = 0; i < data.length; i++)
        {
            current = data[i];
            if(current.containsLocation(location))
            {
                return current;
            }
        }
        return null;
    }

    /**
     * Resolves the Deelgebied on the location.
     *
     * @param location The location where the Deelgebied can be resolved on.
     * @return The Deelgebied resolved from the location, returns null if none.
     */
    public static Deelgebied resolveOnLocation(Location location)
    {
        waitUntilInitzialized();
        Deelgebied[] data = all();
        Deelgebied current;

        for(int i = 0; i < data.length; i++)
        {
            current = data[i];
            if(current.containsLocation(location))
            {
                return current;
            }
        }
        return null;
    }
    public static synchronized void waitUntilInitzialized(){
        /*
        while (!deelgebiedenInitialized){
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }*/
    }
    public static Deelgebied parse(String name) {
        waitUntilInitzialized();
        switch (name.toLowerCase()){
            case "alpha":
            case "Alpha":
            case "a":
            case "A":
                return Alpha;
            case "bravo":
            case "Bravo":
            case "b":
            case "B":
                return Bravo;
            case "charlie":
            case "Charlie":
            case "c":
            case "C":
                return Charlie;
            case "delta":
            case "Delta":
            case "d":
            case "D":
                return Delta;
            case "echo":
            case "Echo":
            case "e":
            case "E":
                return Echo;
            case "foxtrot":
            case "Foxtrot":
            case "f":
            case "F":
                return Foxtrot;
            case "xray":
            case "Xray":
            case "X-ray":
            case "x-ray":
            case "x":
            case "X":
                return Xray;
            default:
                return null;
        }
    }
    public interface OnInitialized {
        void onInitialized(Deelgebied deelgebied);
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/deelgebied/Deelgebied.kt
package nl.rsdt.japp.jotial.maps.deelgebied

import android.content.SharedPreferences
import android.content.res.Resources
import android.graphics.Color
import android.location.Location
import android.util.Log
import com.google.android.gms.maps.model.LatLng
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.maps.android.PolyUtil
import nl.rsdt.japp.R
import nl.rsdt.japp.application.JappPreferences
import nl.rsdt.japp.jotial.data.structures.area348.MetaColorInfo
import nl.rsdt.japp.jotial.maps.kml.KmlDeelgebied
import nl.rsdt.japp.jotial.maps.kml.KmlFile
import nl.rsdt.japp.jotial.maps.kml.KmlReader
import org.acra.ktx.sendWithAcra
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*

/**
 * @author Dingenis Sieger Sinke
 * @version 1.0
 * @since 13-7-2016
 * Class for the Deelgebieden,
 * Deelgebied.initialize(Resources) must be called in order for this class to function.
 */
class Deelgebied
/**
 * Initializes a new instance of Deelgebied.
 *
 * @param name The name of the Deelgebied.
 */
private constructor(
        /**
         * The name of this Deelgebied.
         */
        /**
         * Gets the name of the Deelgebied.
         */
        val name: String, colorName: String) : SharedPreferences.OnSharedPreferenceChangeListener {
    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        val team = this.name.toLowerCase()[0].toString()
        when (key){
            JappPreferences.AREAS_COLOR + team ->{
                changeColor(JappPreferences.getColorName(team))
                for (onColorChangListener in onColorChangeListeners ){
                    onColorChangListener.onColorChange(this)
                }
            }
        }
    }

    /**
     * The hunt drawable associated with this Deelgebied.
     */
    lateinit var dgColor: MetaColorInfo.ColorNameInfo.DeelgebiedColor
    /**
     * Gets the hunt drawable associated with this Deelgebied.
     */
    var drawableHunt: Int = 0
        private set

    /**
     * The spot drawable associated with this Deelgebied.
     */
    /**
     * Gets the spot drawable associated with this Deelgebied.
     */
    var drawableSpot: Int = 0
        private set

    /**
     * The color of this Deelgebied.
     */
    /**
     * Gets the color of the Deelgebied.
     */
    var color: Int = 0
        private set

    /**
     * The list of coordinates that is the area of the Deelgebied.
     */
    /**
     * Gets the list of coordinates that is the area of the Deelgebied.
     */
    var coordinates = ArrayList<LatLng>()
        private set
    private val onColorChangeListeners = LinkedList<OnColorChangeListener>()
    fun addListener(onColorChangeListener: OnColorChangeListener) {
        onColorChangeListeners.add(onColorChangeListener)
    }

    fun removeListener(onColorChangeListener: OnColorChangeListener) {
        onColorChangeListeners.remove(onColorChangeListener)
    }

    /**
     * Gets the color of the Deelgebied, with a given alpha.
     *
     * @param alpha The alpha from 0 - 255.
     * @return The color of the Deelgebied with a given alpha.
     */
    fun alphaled(alpha: Int): Int {
        return Color.argb(alpha, Color.red(color), Color.green(color), Color.blue(color))
    }

    fun getDeelgebiedAsync(onInitialized: OnInitialized) {
        if (deelgebiedenInitialized) {
            onInitialized.onInitialized(this)
        } else {
            if (!onInitializedList.containsKey(this.name)) onInitializedList[this.name] = LinkedList()
            onInitializedList[this.name]!!.add(onInitialized)
        }
    }

    init {
        changeColor(colorName)
    }

    fun changeColor(colorName: String){
        this.dgColor = MetaColorInfo.ColorNameInfo.DeelgebiedColor.valueOf(colorName)
        when (this.dgColor) {

            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Groen -> {
                this.drawableHunt = R.drawable.vos_groen_4
                this.drawableSpot = R.drawable.vos_groen_3
                this.color = Color.argb(255, 0, 255, 0)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Rood -> {
                this.drawableHunt = R.drawable.vos_rood_4
                this.drawableSpot = R.drawable.vos_rood_3
                this.color = Color.argb(255, 255, 0, 0)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Paars -> {
                this.drawableHunt = R.drawable.vos_paars_4
                this.drawableSpot = R.drawable.vos_paars_3
                this.color = Color.argb(255, 255, 0, 255)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Oranje -> {
                this.drawableHunt = R.drawable.vos_oranje_4
                this.drawableSpot = R.drawable.vos_oranje_3
                this.color = Color.argb(255, 255, 162, 0)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Blauw -> {
                this.drawableHunt = R.drawable.vos_blauw_4
                this.drawableSpot = R.drawable.vos_blauw_3
                this.color = Color.argb(255, 0, 0, 255)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Onbekend, MetaColorInfo.ColorNameInfo.DeelgebiedColor.Zwart -> {
                this.drawableHunt = R.drawable.vos_zwart_4
                this.drawableSpot = R.drawable.vos_zwart_3
                this.color = Color.argb(255, 0, 0, 0)
            }
            MetaColorInfo.ColorNameInfo.DeelgebiedColor.Turquoise -> {
                this.drawableHunt = R.drawable.vos_turquoise_4
                this.drawableSpot = R.drawable.vos_turquoise_3
                this.color = Color.argb(255, 0, 255, 255)
            }
            else -> {
                this.drawableHunt = R.drawable.vos_zwart_4
                this.drawableSpot = R.drawable.vos_zwart_3
                this.color = Color.argb(255, 0, 0, 0)
            }
        }
    }
    /**
     * Checks if the DeelgebiedData contains a location.
     *
     * @param location The location to check.
     * @return Value indicating if the location is within the DeelgebiedData.
     */
    fun containsLocation(location: LatLng): Boolean {
        return PolyUtil.containsLocation(location, coordinates, false)
    }

    /**
     * Checks if the DeelgebiedData contains a location.
     *
     * @param location The location to check.
     * @return Value indicating if the location is within the DeelgebiedData.
     */
    fun containsLocation(location: Location): Boolean {
        return PolyUtil.containsLocation(LatLng(location.latitude, location.longitude), coordinates, false)
    }

    interface OnInitialized {
        fun onInitialized(deelgebied: Deelgebied)
    }

    companion object {

        /**
         * Defines the Alpha Deelgebied.
         */
        private val Alpha = Deelgebied("alpha", JappPreferences.getColorName("a")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Groen.toString())

        /**
         * Defines the Bravo Deelgebied.
         */
        private val Bravo = Deelgebied("bravo", JappPreferences.getColorName("b")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Oranje.toString())

        /**
         * Defines the Charlie Deelgebied.
         */
        private val Charlie = Deelgebied("charlie", JappPreferences.getColorName("c")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Rood.toString())

        /**
         * Defines the Delta Deelgebied.
         */
        private val Delta = Deelgebied("delta", JappPreferences.getColorName("d")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Turquoise.toString())

        /**
         * Defines the Echo Deelgebied.
         */
        private val Echo = Deelgebied("echo", JappPreferences.getColorName("e")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Blauw.toString())

        /**
         * Defines the Foxtrot Deelgebied.
         */

        private val Foxtrot = Deelgebied("foxtrot", JappPreferences.getColorName("f") ?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Paars.toString())

        /**
         * Defines the Xray Deelgebied.
         */
        val Xray = Deelgebied("xray", JappPreferences.getColorName("x")?: MetaColorInfo.ColorNameInfo.DeelgebiedColor.Zwart.toString())
        private val TAG = "Deelgebied"

        /**
         * Gets a array of all the Deelgebieden.
         */
        fun all(): Array<Deelgebied> {
            return arrayOf(Alpha, Bravo, Charlie, Delta, Echo, Foxtrot, Xray)
        }

        @Volatile
        var deelgebiedenInitialized = false
        private val onInitializedList = HashMap<String, LinkedList<OnInitialized>>()


        /**
         * Initializes the Deelgebieden, loading their coordinates from the Resources.
         */
        @Synchronized
        fun initialize(resources: Resources) {
            KmlReader.parseFromMeta(object : KmlReader.Callback {
                override fun onException(e: Throwable) {
                    e.sendWithAcra()
                    Log.e(Deelgebied.TAG, e.toString())
                    val gebieden = all()
                    var current: Deelgebied
                    for (g in gebieden.indices) {
                        current = gebieden[g]

                        val stream: InputStream?
                        when (current.name) {
                            "alpha" -> stream = resources.openRawResource(R.raw.alpha)
                            "bravo" -> stream = resources.openRawResource(R.raw.bravo)
                            "charlie" -> stream = resources.openRawResource(R.raw.charlie)
                            "delta" -> stream = resources.openRawResource(R.raw.delta)
                            "echo" -> stream = resources.openRawResource(R.raw.echo)
                            //case "foxtrot":
                            //stream = resources.openRawResource(R.raw.foxtrot);
                            //break;
                            else -> {
                                stream = null
                                Log.i("Deelgebied", "No polygon data was found for " + current.name)
                            }
                        }

                        if (stream != null) {
                            val r = BufferedReader(InputStreamReader(stream))
                            val total = StringBuilder()
                            var line: String? = r.readLine()
                            try {
                                while (line  != null) {
                                    total.append(line).append('\n')
                                    line = r.readLine()
                                }
                            } catch (err: IOException) {
                                Log.e("Deelgebied", "Error occurred while reading stream", err)
                                e.sendWithAcra()
                            }

                            val data = total.toString()

                            current.coordinates = Gson().fromJson(data, object : TypeToken<ArrayList<LatLng>>() {

                            }.type)
                        }
                    }
                    deelgebiedenInitialized = true
                    for ((key, value) in onInitializedList) {
                        for (oi in value) {
                            oi.onInitialized(Objects.requireNonNull<Deelgebied>(Deelgebied.parse(key)))
                        }
                    }
                }

                override fun onSucces(kml: KmlFile) {
                    val gebieden = all()
                    var current: Deelgebied
                    for (g in gebieden.indices) {
                        current = gebieden[g]

                        val kmldg: KmlDeelgebied?
                        when (current.name) {
                            "alpha" -> kmldg = kml.alpha
                            "bravo" -> kmldg = kml.bravo
                            "charlie" -> kmldg = kml.charlie
                            "delta" -> kmldg = kml.delta
                            "echo" -> kmldg = kml.echo
                            "foxtrot" -> kmldg = kml.foxtrot
                            else -> kmldg = null
                        }

                        if (kmldg != null) {

                            val tmp = ArrayList<LatLng>(kmldg.boundry.size)
                            for (i in 0 until kmldg.boundry.size) {
                                val kmldgLoc = kmldg.boundry[i]
                                val tmploc = LatLng(kmldgLoc.lat?:0.0, kmldgLoc.lon?:0.0)
                                while (tmp.size - 1 < i) {
                                    tmp.add(LatLng(0.0,0.0))
                                }
                                tmp[i] = tmploc
                            }
                            current.coordinates = tmp
                        } else {
                            Log.i("Deelgebied", "No polygon data was found for " + current.name)
                        }
                    }
                    deelgebiedenInitialized = true
                    for ((key, value) in onInitializedList) {
                        for (oi in value) {
                            parse(key)?.also { oi.onInitialized(it) }
                        }
                    }
                }
            })
            waitUntilInitzialized()
        }

        /**
         * Resolves the Deelgebied on the location.
         *
         * @param location The location where the Deelgebied can be resolved on.
         * @return The Deelgebied resolved from the location, returns null if none.
         */
        fun resolveOnLocation(location: LatLng): Deelgebied? {
            waitUntilInitzialized()
            val data = all()
            var current: Deelgebied

            for (i in data.indices) {
                current = data[i]
                if (current.containsLocation(location)) {
                    return current
                }
            }
            return null
        }

        /**
         * Resolves the Deelgebied on the location.
         *
         * @param location The location where the Deelgebied can be resolved on.
         * @return The Deelgebied resolved from the location, returns null if none.
         */
        fun resolveOnLocation(location: Location): Deelgebied? {
            waitUntilInitzialized()
            val data = all()
            var current: Deelgebied

            for (i in data.indices) {
                current = data[i]
                if (current.containsLocation(location)) {
                    return current
                }
            }
            return null
        }

        @Synchronized
        fun waitUntilInitzialized() {
            /*
        while (!deelgebiedenInitialized){
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {

                e.printStackTrace();
            }
        }*/
        }

        fun parse(name: String?): Deelgebied {

            waitUntilInitzialized()
            return if (name == null){
                Xray
            }else{
                when (name.lowercase(Locale.ROOT)) {
                    "alpha", "Alpha", "a", "A" -> Alpha
                    "bravo", "Bravo", "b", "B" -> Bravo
                    "charlie", "Charlie", "c", "C" -> Charlie
                    "delta", "Delta", "d", "D" -> Delta
                    "echo", "Echo", "e", "E" -> Echo
                    "foxtrot", "Foxtrot", "f", "F" -> Foxtrot
                    "xray", "Xray", "X-ray", "x-ray", "x", "X" -> Xray
                    else -> Xray
                }
            }
        }
    }
    interface OnColorChangeListener{
        fun onColorChange(deelgebied: Deelgebied)
    }
}
